libbpf-tools/funclatency.bpf.c

```diff
@@ -0,0 +1,84 @@
+// SPDX-License-Identifier: GPL-2.0
+/* Copyright (c) 2021 Google LLC. */
+#include "vmlinux.h"
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_tracing.h>
+#include "funclatency.h"
+#include "bits.bpf.h"
+
+const volatile pid_t targ_tgid;
+const volatile int units;
+
+/* key: pid.  value: start time */
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__uint(max_entries, MAX_PIDS);
+	__type(key, u32);
+	__type(value, u64);
+} starts SEC(".maps");
+
+/* key: hist slot idx.  value: count */
+struct {
+	__uint(type, BPF_MAP_TYPE_PERCPU_ARRAY);
+	__uint(max_entries, MAX_SLOTS);
+	__type(key, u32);
+	__type(value, u64);
+} hist SEC(".maps");
+
+SEC("kprobe/dummy_kprobe")
+int BPF_KPROBE(dummy_kprobe)
+{
+	u64 id = bpf_get_current_pid_tgid();
+	u32 tgid = id >> 32;
+
+	u64 nsec;
+
+	if (targ_tgid && targ_tgid != tgid)
+		return 0;
+	nsec = bpf_ktime_get_ns();
+	bpf_map_update_elem(&starts, &id, &nsec, BPF_ANY);
+
+	return 0;
+}
+
+static void update_hist(u64 nsec)
+{
+	u64 slot, val, *count;
+
+	val = nsec;
+	switch (units) {
+	case USEC:
+		val /= 1000;
+		break;
+	case MSEC:
+		val /= 1000000;
+		break;
+	}
+
+	slot = log2l(val);
+	if (slot >= MAX_SLOTS)
+		slot = MAX_SLOTS - 1;
+	count = bpf_map_lookup_elem(&hist, &slot);
```

Comment:
again, map expects u32, but you are passing u64

Reply:
agreed that should be a u32.  FWIW, the reason i had `u64 slot` was from earlier prototyping on a global array:

`__sync_fetch_and_add(&hist[slot], 1);`

for that, slot needs to be a u64, otherwise the program was failing verification.